### PR TITLE
[Fluid] FS strategy clean up

### PR DESCRIPTION
**Pull request by rubenzorrilla**
As suggested by @RiccardoRossi this PR brings the FS strategy to the current standards. This includes:
- Remove the `Solve` method in favor of the `InitializeSolutionStep`, `SolveSolutionStep` and `FinalizeSolutionStep` ones.
- 2 new constructors with a calculate reactions flag are added. The old ones are kept for the sake of backwards compatibility but throw a deprecation warning. This allows to use the `SolveSolutionStep`method in the base `FluidSolver` as the reactions are now calculated inside the `SolveSolutionStep` method of the c++ strategy. In consequence, the Python exposure of the `CalculateReactions` method is also deprecated.
- All the loops using the `OpenMPUtils` are rewritten using OpenMP specifiers.
- A minor clean up of the includes.
- Information output using the Kratos macros.

### CODE REVIEW

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (adityaghantasala) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h:200]: should'nt there be an error thrown here ?

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (adityaghantasala) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h:207]: Also here ?

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (philbucher) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h:200]: this should not even be done in the Strategy IMO
=> #6754

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (adityaghantasala) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]:
```suggestion
        const int number_of_elements = static_cast<int>(r_model_part.NumberOfElements());
        for (int i_elem = 0; i_elem < number_of_elements; ++i_elem) {
```

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (adityaghantasala) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: Also here we can define number_of_nodes outside the for statement once and use it.

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (adityaghantasala) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: you can define the ZeroVector(3) outside for loop as const and use it here.

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (adityaghantasala) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: `static_cast<int>(rModelPart.NumberOfNodes())` as const before the loop

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (adityaghantasala) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: `static_cast<int>(rModelPart.NumberOfNodes())` as const before the loop

File: applications/FluidDynamicsApplication/trilinos_extension/custom_python/add_trilinos_strategies_to_python.cpp
Comment (adityaghantasala) [inline applications/FluidDynamicsApplication/trilinos_extension/custom_python/add_trilinos_strategies_to_python.cpp]: Why depreciating and not completely removing this ?

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (adityaghantasala) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h:200]: perfect .. then I vote for removing this part !

**Comment by adityaghantasala** [comment]
@rubenzorrilla  a very helpful and nice cleanup thank you very much. I think at some point we should also make necessary changes to remove the FractionalStepSettings and related functions/classes. 

- There are a couple of functions ( including constructors) which are depreciated. I think we can completely remove those instead of now depreciating and then later removing them.

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h:200]: > should'nt there be an error thrown here ?

The error is thrown outside. In any case I'd wait for #6754 to be merged so we can remove it.

File: applications/FluidDynamicsApplication/trilinos_extension/custom_python/add_trilinos_strategies_to_python.cpp
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/trilinos_extension/custom_python/add_trilinos_strategies_to_python.cpp]: Backwards compatibility.

**Comment by rubenzorrilla** [comment]
> @rubenzorrilla a very helpful and nice cleanup thank you very much. I think at some point we should also make necessary changes to remove the FractionalStepSettings and related functions/classes.

I thought a bit about it, and I actually like this design. It makes the construction of the strategy pretty clean. In any case we can OFC discuss any alternative design.

> 
>     * There are a couple of functions ( including constructors) which are depreciated. I think we can completely remove those instead of now depreciating and then later removing them.

Instead of directly remove them, I'd prefer to deprecate them and remove them after some weeks... I'm afraid these might be used in some forks.

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h:207]: (same comment above)

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: I decided to use the `Zero()` method of the variables to avoid allocating the array.

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (philbucher) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: but the velocity can still not be converged right?
So unless I am using predictor corrector it will always tell that it converged?

Not sure if this is a good idea

But maybe I misunderstood

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (philbucher) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: better use variable-utils for this

I guess I don't have to point you to the functions since you wrote them :D

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (philbucher) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]:
```suggestion
            Vector& BDFcoeffs = r_process_info[BDF_COEFFICIENTS];
```

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (philbucher) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h:642]: probably not for this PR, but do those need to be historical?

**Review by philbucher — COMMENTED** [review]
nice work, minor comments

@mpentek maybe you want to test this?

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h:642]: I don't think so. In any case, this is a delicate change that would require a PR on its own.

Comment (philbucher) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h:642]: > this is a delicate change that would require a PR on its own.

absolutely!

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: The velocity convergence is check in the `SolveStep`, which throws a warning in case the momentum splitting does not convergence.

This was already there. I think that the original point was that you likely do not converge in terms of the pressure. As this is "intentionally" done for the sake of efficiency, checking the pressure norm and throwing a warning in the majority of the cases makes no sense at all. We can of course change it, but I'd rather keep it as it is.

Comment (adityaghantasala) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: We can return the convergence state from `SolveStep` . that is the convergence of velocity iterations.

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: :laughing:

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (adityaghantasala) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h:642]: There are also non historical vars (of the same names) used in the strategy :-)

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (philbucher) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: > We can return the convergence state from `SolveStep` . that is the convergence of velocity iterations.

that sounds like the logical thing to do for me

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: Done.

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: OK. We can change the return type of the `SolveStep` to do this.

Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: Done. Thanks guys!

**Comment by rubenzorrilla** [comment]
I think all concerns are now addressed. Can someone approve @adityaghantasala @philbucher @RiccardoRossi ?

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (philbucher) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h:142]: I see now that you have 4 constructors that essentially do the same

I recommend calling one constructor from the other (possible since C++11) like I did in `ModelPart`, this should reduce code duplication

up to you

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (philbucher) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: I recommend (actually there is no other way which I had to learn the hard way...) creating a const-ref to the processInfo and pass it here
Otherwise this will screw you up when you start using the new element interfaces

Comment (philbucher) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: BTW isn't this a bit weird, using `CalculateLocalSystem` to compute the reactions?
Wouldn't this be sth for `Calculate`?

Comment (philbucher) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: This might be relevant in #5121

**Review by philbucher — COMMENTED** [review]
some more comments otherwise looks good to me 👍

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h:142]: Considering that two constructors are deprecated, and hopefully we will remove them soon, I'd keep it as it is.

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: > I recommend (actually there is no other way which I had to learn the hard way...) creating a const-ref to the processInfo and pass it here
> Otherwise this will screw you up when you start using the new element interfaces

Done. Thanks!

Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: > BTW isn't this a bit weird, using `CalculateLocalSystem` to compute the reactions?
> Wouldn't this be sth for `Calculate`?

The point is that we compute the `REACTION` as minus the residual contribution once the problem has been solved. As the `FractionalStep` element does not implement the RHS-only calculation method, we can only do that by recomputing the local system. I agree this is way inefficient but should be handled in a different PR.

Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h]: > This might be relevant in #5121

mmm... I don't think #5121 influences the `FractionalStepStrategy` as it doesn't require a scheme.

**Comment by rubenzorrilla** [comment]
@philbucher done. Some comments:
1) I renamed `FSStrategy` to `FractionalStepStrategy` as you suggested. I agree this is much more clear now.
2) With regard to the `Parameters`-based constructor, it is not straightforward as the `FractionalStepStrategy` uses a custom container `StrategySettings` to build the velocity and pressure substrategies. I'd rather do it in an further PR that only implements this.

**Review by philbucher — APPROVED** [review]
looks ok to me, thx for the cleanup

CoSim tests run fine

I am approving but maybe you want to wait for the approval of @KratosMultiphysics/fluid-dynamics

**Comment by rubenzorrilla** [comment]
As there are no more comments, I assume we are ready to merge this.

**Pull request merged**
